clouddriver-appengine/src/main/groovy/com/netflix/spinnaker/clouddriver/appengine/provider/agent/AppEngineServerGroupCachingAgent.groovy

```diff
@@ -0,0 +1,438 @@
+            relationships[INSTANCES.ns].addAll(instanceKeys)
+          }
+        }
+      }
+    }
+
+    log.info("Caching ${cachedApplications.size()} applications in ${agentType}")
+    log.info("Caching ${cachedClusters.size()} clusters in ${agentType}")
+    log.info("Caching ${cachedServerGroups.size()} server groups in ${agentType}")
+    log.info("Caching ${cachedLoadBalancers.size()} load balancers in ${agentType}")
+    log.info("Caching ${cachedInstances.size()} instances in ${agentType}")
+
+    new DefaultCacheResult([
+      (APPLICATIONS.ns): cachedApplications.values(),
+      (CLUSTERS.ns): cachedClusters.values(),
+      (SERVER_GROUPS.ns): cachedServerGroups.values(),
+      (LOAD_BALANCERS.ns): cachedLoadBalancers.values(),
+      (INSTANCES.ns): cachedInstances.values(),
+      (ON_DEMAND.ns): onDemandKeep.values()
+    ], [
+      (ON_DEMAND.ns): onDemandEvict
+    ])
+  }
+
+  static void cache(Map<String, List<CacheData>> cacheResults,
+                    String cacheNamespace,
+                    Map<String, CacheData> cacheDataById) {
+    cacheResults[cacheNamespace].each {
+      def existingCacheData = cacheDataById[it.id]
+      if (existingCacheData) {
+        existingCacheData.attributes.putAll(it.attributes)
+        it.relationships.each { String relationshipName, Collection<String> relationships ->
+          existingCacheData.relationships[relationshipName].addAll(relationships)
+        }
+      } else {
+        cacheDataById[it.id] = it
+      }
+    }
+  }
+
+  Map<String, List<Version>> loadServerGroups() {
+    def project = credentials.project
+    def loadBalancers = credentials.appengine.apps().services().list(project).execute().getServices() ?: []
+    BatchRequest batch = credentials.appengine.batch()
+    Map<String, List<Version>> serverGroupsByLoadBalancerName = [:].withDefault { [] }
+
+    loadBalancers.each { loadBalancer ->
+      def loadBalancerName = loadBalancer.getId()
+      def callback = new AppEngineCallback<ListVersionsResponse>()
+        .success { ListVersionsResponse versionsResponse, HttpHeaders responseHeaders ->
+          def versions = versionsResponse.getVersions()
+          if (versions) {
+            serverGroupsByLoadBalancerName[loadBalancerName].addAll(versions)
+          }
+        }
+
+      credentials.appengine.apps().services().versions().list(project, loadBalancerName).queue(batch, callback)
+    }
+
+    batch.execute()
+    serverGroupsByLoadBalancerName
+  }
+
+  Map loadServerGroupAndLoadBalancer(String serverGroupName) {
+    def project = credentials.project
+    def loadBalancers = credentials.appengine.apps().services().list(project).execute().getServices() ?: []
+    BatchRequest batch = credentials.appengine.batch()
+    Service loadBalancer
+    Version serverGroup
+
+    // We don't know where our server group is, so we have to check all of the load balancers.
+    loadBalancers.each { Service lb ->
```

Comment:
If this is the case, I would fill in the relationship data for LB -> SG & SG -> LB in the load balancer caching agent, since you'll be loading all the LBs there anyhow.